airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py

```diff
@@ -0,0 +1,273 @@
+            self._call_credit.consume(credit)
+    
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        self._consume_credit(self.call_credit)
+        yield from super().read_records(sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=stream_slice, stream_state=stream_state)
+    
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        data = response.json()
+        if not data:
+            return
+        for element in data:
+            yield element
+
+
+class IncrementalFreshdeskStream(FreshdeskStream, ABC):
+
+    state_filter = "updated_since"  # Name of filter that corresponds to the state
+
+    @property
+    def cursor_field(self) -> str:
+        return "updated_at"
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        current_stream_state = current_stream_state or {}
+
+        current_stream_state_date = current_stream_state.get("updated_at", self.start_date)
+        if isinstance(current_stream_state_date, str):
+            current_stream_state_date = pendulum.parse(current_stream_state_date)
+        latest_record_date = pendulum.parse(latest_record.get("updated_at")) if latest_record.get("updated_at") else self.start_date
+
+        return {"updated_at": max(current_stream_state_date, latest_record_date).strftime("%Y-%m-%dT%H:%M:%SZ")}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        if "updated_at" in stream_state:
+            params[self.state_filter] = stream_state["updated_at"]
+        else:
+            params[self.state_filter] = self.start_date
+        return params
+
+
+class Agents(FreshdeskStream):
+
+    def path(self, **kwargs) -> str:
+        return "agents"
+
+
+class Companies(FreshdeskStream):
+
+    def path(self, **kwargs) -> str:
+        return "companies"
+
+
+class Contacts(IncrementalFreshdeskStream):
+    state_filter = "_updated_since"
+
+    def path(self, **kwargs) -> str:
+        return "contacts"
+
+
+class Groups(FreshdeskStream):
+
+    def path(self, **kwargs) -> str:
+        return "groups"
+
+
+class Roles(FreshdeskStream):
+
+    def path(self, **kwargs) -> str:
+        return "roles"
+
+
+class Skills(FreshdeskStream):
+
+    def path(self, **kwargs) -> str:
+        return "skills"
+
+
+class TimeEntries(FreshdeskStream):
+
+    def path(self, **kwargs) -> str:
+        return "time_entries"
+
+
+class Tickets(IncrementalFreshdeskStream):
```

Comment:
Given that **_Tickets_** will be retrieved in ascending order maybe we can set a **_state_checkpoint_interval = 100_** in order to persist state on every 100 records and avoid sending previously sent records if the replication crashes midway?

https://github.com/airbytehq/airbyte/blob/master/docs/connector-development/cdk-python/incremental-stream.md#interval-based-checkpointing

Reply:
Awesome. I added this in [4e2f437](https://github.com/airbytehq/airbyte/pull/12334/commits/4e2f4374641f906dfc9b9c69a77f02847186ef65). Thank you for the suggestion!

Reply:
I suggest you set `state_checkpoint_interval` on all the Incremental streams, or you can implement `stream_slices`. If you implement stream slicing the state will be checkpointed once a slice was properly consumed.

Reply:
Added in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f).